src/classes/frDonation.cls

```diff
@@ -62,76 +62,180 @@ public class frDonation extends frModel implements frSyncable {
-                if (campaigns.size() > 0) {
-                    o.CampaignId = campaigns[0].Id;
+            
+            Boolean campaignMappingDisabled = request.containsKey(META_CAMPAIGN_MAPPING_KEY) ?
+                Boolean.valueOf(request.get(META_CAMPAIGN_MAPPING_KEY)): true;
+            if(!campaignMappingDisabled) {
+                String campaignId = String.valueOf(request.get('campaignGoalId'));
+                if (String.isNotBlank(campaignId)) {
+                    List<Campaign> campaigns = [SELECT Id FROM Campaign WHERE fr_ID__c = :campaignId];
+                    if (campaigns.size() > 0) {
+                        gt.put('CampaignId', campaigns[0].Id);
+                        //gt.CampaignId = campaigns[0].Id;
+                    } else {
+                        frUtil.logRelationshipError(getFrType(), funraiseId, 
+                                                    frUtil.Entity.CAMPAIGN, campaignId);
+                    }
+                }
+            }
+            
+            String subscriptionId = String.valueOf(request.get('subscriptionId'));
+            if (String.isNotBlank(subscriptionId)) {
+                List<SObject> subscriptions = Database.query('SELECT Id FROM GiftCommitment WHERE fr_ID__c = :subscriptionId');
+                if (subscriptions.size() > 0) {
+                    //gt.GiftCommitmentId = subscriptions.get(0).Id;
+                    gt.put('GiftCommitmentId', subscriptions.get(0).get('Id'));
                 } else {
                     frUtil.logRelationshipError(getFrType(), funraiseId, 
-                                                frUtil.Entity.CAMPAIGN, campaignId);
+                                                frUtil.Entity.SUBSCRIPTION, subscriptionId);
                 }
             }
-        }
-        
-        String subscriptionId = String.valueOf(request.get('subscriptionId'));
-        if (String.isNotBlank(subscriptionId)) {
-            List<Subscription__c> subscriptions = [SELECT Id FROM Subscription__c WHERE fr_ID__c = :subscriptionId];
-            if (subscriptions.size() > 0) {
-                o.Subscription__c = subscriptions.get(0).Id;
-            } else {
-                frUtil.logRelationshipError(getFrType(), funraiseId, 
-                                            frUtil.Entity.SUBSCRIPTION, subscriptionId);
+            
+            Boolean isPledge = Boolean.valueOf(request.get('pledge'));
+            List<SObject> alreadyMappedToPledgeList= Database.query('SELECT id FROM GiftTransaction WHERE fr_Id__c = :funraiseId AND Funraise_Pledge__c != null');
+            Boolean alreadyMappedToPledge = alreadyMappedToPledgeList.size() > 0;
+            if(!isPledge && !alreadyMappedToPledge) {
+                Pledge__c pledge = frPledge.findActive((Id) gt.get('DonorId'), (Date) gt.get('AcknowledgementDate'));
+                NPC_Setting__c npcSetting = NPC_Setting__c.getOrgDefaults();
```

Comment:
Can you explain more on the purpose of this new custom setting that seems to be storing a global Pledge value?  I think it would be unexpected for every transaction to count towards the same pledge

Reply:
The custom setting value holds the name of the field to populate for pledge since we cant create the lookup due to it being an NPC object

Reply:
Hm this would mean that in NPC orgs we'll have to instruct them that a custom lookup field and custom setting needs to be created and then they will have to enter that in the custom setting?

Can we change this so that the field name will be something that we hardcode, and log a funraise_Error__c if they have not created the lookup?